### PR TITLE
fix(dynamodb): cannot use attribute as key in a GSI, non-key in another

### LINKED ISSUE #4398: Cannot use key attribute in one DynamoDB index as non-key attribute in another index
I'm using the CDK (in Python) to create global secondary indexes in my DynamoDB table, and one secondary index has a key that's used as a non-key attribute in another index. I have no problems doing this manually in the DynamoDB console, but when I use CDK, I get an error because CDK seems to check the union of key attributes across all indexes against the union of non-key attributes across all indexes. I traced the problem to [this commit](https://github.com/aws/aws-cdk/pull/760), specifically these lines:

* https://github.com/aws/aws-cdk/pull/760/files#diff-412803b83e4ab5b404012c6e96e18aebR237-R244
* https://github.com/aws/aws-cdk/pull/760/files#diff-412803b83e4ab5b404012c6e96e18aebR285
* https://github.com/aws/aws-cdk/pull/760/files#diff-412803b83e4ab5b404012c6e96e18aebR424-R426

Was this behavior intentional? This seems to restrict what one can do with CDK vs DynamoDB directly.

### Reproduction Steps

In the following snippet, `bar` is used as a non-key attribute in Index1 and as a key attribute in Index2.

```python
my_table.add_global_secondary_index(
    index_name = "Index1",
    partition_key = aws_dynamodb.Attribute(
        name = "foo",
        type = aws_dynamodb.AttributeType.STRING
    ),
    projection_type = aws_dynamodb.ProjectionType.INCLUDE,
    non_key_attributes = ["bar"] 
)
my_table.add_global_secondary_index(
    index_name = "Index2",
    partition_key = aws_dynamodb.Attribute(
        name = "baz",
        type = aws_dynamodb.AttributeType.STRING
    ),
    sort_key = aws_dynamodb.Attribute(
        name = "bar",
        type = aws_dynamodb.AttributeType.STRING
    ),
    projection_type = aws_dynamodb.ProjectionType.INCLUDE,
    non_key_attributes = ["blah"] 
)
```

### Error Log

`Error: a key attribute, bar, is part of a list of non-key attributes, bar,blah, which is not allowed since all key attributes are added automatically and this configuration causes stack creation failure`

### Environment

  - **CLI Version      :**
  - **Framework Version:**
  - **OS               :**  Linux
  - **Language         :**  Python

### Other

This could probably be fixed by changing the `addGlobalSecondaryIndex` method in `Table` so the key and non-key attributes are only compared per index. 

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by RomainMuller**
### Commit Message

Removes the restriction that a key attribute cannot be used as a non-key
attribute in a distinct secondary index. This was excessively
restricting what can be done, as new integration tests demonstrated that
there is no problem in deploying such a configuration.

Also, switches from `Array`s to `Set`s to represent the sets of names that
must be unique.

Fixes #4398

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-dynamodb/test/dynamodb.test.ts [hunk 1]
```diff
@@ -0,0 +1,1758 @@
...
+  AttributeType,
+  BillingMode,
+  GlobalSecondaryIndexProps,
+  LocalSecondaryIndexProps,
+  ProjectionType,
+  StreamViewType,
+  Table,
+} from '../lib';
+
+// tslint:disable:object-literal-key-quotes
```
Comment (nija-at) [inline packages/@aws-cdk/aws-dynamodb/test/dynamodb.test.ts]: Looks like this can be dropped.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts [hunk 2]
```diff
@@ -908,21 +908,13 @@ export class Table extends TableBase {
...
-    this.nonKeyAttributes.push(...nonKeyAttributes);
-
-    // throw error if key attribute is part of non-key attributes
-    this.attributeDefinitions.forEach(keyAttribute => {
-      if (typeof keyAttribute.attributeName === 'string' && this.nonKeyAttributes.includes(keyAttribute.attributeName)) {
-        throw new Error(`a key attribute, ${keyAttribute.attributeName}, is part of a list of non-key attributes, ${this.nonKeyAttributes}` +
-          ', which is not allowed since all key attributes are added automatically and this configuration causes stack creation failure');
-      }
-    });
+    nonKeyAttributes.forEach(att => this.nonKeyAttributes.add(att));
```
Comment (nija-at) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts]: can you confirm you've added unit test to this effect?

Comment (RomainMuller) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: I literally only changed the style around here (was array and using push(...) now is Set and add), not the feature.

File: packages/@aws-cdk/aws-dynamodb/test/dynamodb.test.ts (hunk 1 above)
Comment (RomainMuller) [reply packages/@aws-cdk/aws-dynamodb/test/dynamodb.test.ts:17]: Turns out it can't 🤷🏻‍♂️

**Pull request merged**